oprf/oprf.go

```diff
@@ -0,0 +1,238 @@
+	ErrUnsupportedGroup = errors.New("the chosen group is not supported")
+)
+
+// BlindToken corresponds to a token that has been blinded.
+// Internally, it is a serialized Element.
+type BlindToken []byte
+
+// IssuedToken corresponds to a token that has been issued.
+// Internally, it is a serialized Element.
+type IssuedToken []byte
+
+// Token is the object issuance of the protocol.
+type Token struct {
+	data  []byte
+	blind *group.Scalar
+}
+
+// Evaluation corresponds to the evaluation over a token.
+type Evaluation struct {
+	element []byte
+}
+
+// KeyPair is an struct containing a public and private key.
+type KeyPair struct {
+	PubK  *group.Element
+	PrivK *group.Scalar
+}
+
+// Client is a representation of a Client during protocol execution.
+type Client struct {
+	suite *group.Ciphersuite
+	ctx   string
+}
+
+// ClientContext implements the functionality of a Client.
+type ClientContext interface {
+	// Blind generates a token and blinded data.
+	Blind(in []byte) (*Token, *BlindToken, error)
+
+	// Unblind unblinds the server response.
+	Unblind(t *Token, e *Evaluation) (IssuedToken, error)
+
+	// Finalize outputs a byte array that corresponds to its input.
+	Finalize(t *Token, issuedT IssuedToken, info []byte) []byte
+}
+
+// Server is a representation of a Server during protocol execution.
+type Server struct {
+	suite *group.Ciphersuite
+	ctx   string
+	K     *KeyPair
+}
+
+// ServerContext implements the functionality of a Server.
+// TODO: add FullEvaluate
+type ServerContext interface {
+	// Evaluate evaluates the token.
+	Evaluate(b BlindToken) *Evaluation
+}
+
+func generateCtx(suiteID uint16) string {
+	ctx := OPRFMode + fmt.Sprintf("%x", suiteID)
+
+	return ctx
+}
+
+func generateKeys(suite *group.Ciphersuite) (*KeyPair, error) {
+	privK, err := suite.RandomScalar()
+	if err != nil {
+		return nil, err
+	}
+
+	pubK := suite.ScalarMultBase(privK)
+
+	return &KeyPair{pubK, privK}, nil
+}
+
+func suiteFromID(suiteID uint16) (*group.Ciphersuite, error) {
+	var err error
+	var suite *group.Ciphersuite
+
+	// TODO: add other suites
+	switch suiteID {
+	case OPRFP256:
+		suite, err = group.NewSuite("P-256")
+	case OPRFP384:
+		suite, err = group.NewSuite("P-384")
+	case OPRFP521:
+		suite, err = group.NewSuite("P-521")
+	default:
+		return suite, ErrUnsupportedGroup
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return suite, nil
+}
+
+// NewServer creates a new instantiation of a Server.
+func NewServer(suiteID uint16) (*Server, error) {
+	suite, err := suiteFromID(suiteID)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx := generateCtx(suiteID)
+
+	keyPair, err := generateKeys(suite)
+	if err != nil {
+		return nil, err
+	}
+
+	server := &Server{}
+	server.suite = suite
+	server.ctx = ctx
+	server.K = keyPair
+
+	return server, nil
+}
+
+// Evaluate creates an evaluation of the blided token.
+func (s *Server) Evaluate(b BlindToken) *Evaluation {
+	p := group.NewElement(s.suite)
+	p.Deserialize(b)
```

Comment:
What if deserialization fails? Should we not return an error?

Reply:
How will it fail?

Reply:
Fixed now.